pkg/cmd/mirror/mirror.go

```diff
@@ -230,21 +459,70 @@ def go_deps():
 `, repoName, buildFileProtoModeForRepo(repoName))
 
 		dumpBuildNamingConventionArgsForRepo(repoName)
+		expectedURL := formatURL(replaced.Path, replaced.Version)
 		fmt.Printf("        importpath = \"%s\",\n", mod.Path)
 		if err := dumpPatchArgsForRepo(repoName); err != nil {
 			return err
 		}
-		d := downloaded[replaced.Path]
-		if mod.Replace != nil {
-			fmt.Printf("        replace = \"%s\",\n", replaced.Path)
-		}
-		fmt.Printf(`        sum = "%s",
+		oldMirror, ok := existingMirrors[repoName]
+		if ok && oldMirror.url == expectedURL {
+			// The URL matches, so just reuse the old mirror.
+			fmt.Printf(`        sha256 = "%s",
+        strip_prefix = "%s@%s",
+        urls = [
+            "%s",
+        ],
+`, oldMirror.sha256, replaced.Path, replaced.Version, oldMirror.url)
+		} else if canMirror() {
+			// We'll have to mirror our copy of the zip ourselves.
+			d := downloaded[replaced.Path]
+			sha, err := getSha256OfFile(d.Zip)
+			if err != nil {
+				return fmt.Errorf("could not get zip for %v: %w", *replaced, err)
+			}
+			fmt.Printf(`        sha256 = "%s",
+        strip_prefix = "%s@%s",
+        urls = [
+            "%s",
+        ],
+`, sha, replaced.Path, replaced.Version, expectedURL)
+			waitingTasks++
+			go uploadFile(client, d.Zip, formatSubURL(replaced.Path, replaced.Version), uploadTasks)
+		} else {
+			// We don't have a mirror and can't upload one, so just
+			// have Gazelle pull the repo for us.
+			d := downloaded[replaced.Path]
+			if mod.Replace != nil {
+				fmt.Printf("        replace = \"%s\",\n", replaced.Path)
+			}
+			// Note: `build/teamcity-check-genfiles.sh` checks for
+			// the presence of the "TODO: mirror this repo" comment.
+			// Don't update this comment without also updating the
+			// script.
+			fmt.Printf(`        # TODO: mirror this repo (to fix, run `+"`./dev generate bazel --mirror`"+`
+        sum = "%s",
         version = "%s",
-    )
 `, d.Sum, d.Version)
+		}
+		fmt.Println("    )")
 	}
 
-	return nil
+	// Wait for uploads to complete.
+	for waitingTasks > 0 {
```

Comment:
I wonder if we can simplify this logic and use `errgroup`, see https://pkg.go.dev/golang.org/x/sync/errgroup#example-Group-JustErrors

Reply:
Done.